### PR TITLE
trivial: remove some legacy code, make global a `unique_ptr`, simplify `get_next_work` test, make UBSan CI halt on error (misc. changes)

### PR DESCRIPTION
**Pull request by kwvg**
## Additional Information

Collection of miscellaneous changes collected from work on earlier pull requests that don't fit into pull requests in the immediate future but are nonetheless useful.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas **(note: N/A)**
- [x] I have added or updated relevant unit/integration/functional/e2e tests 
- [x] I have made corresponding changes to the documentation **(note: N/A)**
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

**Review by UdjinM6 — APPROVED** [review]
utACK bd4206fc9eafa3c2dffb2cce0acbf5d610d9d6b9

File: src/test/pow_tests.cpp [hunk 1]
```diff
@@ -11,112 +11,40 @@
...
-    blockIndexPrev22.nTime = 1408728332;  // Block #123434
-    blockIndexPrev22.nBits = 0x1b10e09e;
-    blockIndexPrev21.pprev = &blockIndexPrev22;
-    CBlockIndex blockIndexPrev23 = CBlockIndex();
-    blockIndexPrev23.nTime = 1408728124;  // Block #123433
-    blockIndexPrev23.nBits = 0x1b104be1;
-    blockIndexPrev22.pprev = &blockIndexPrev23;
+    // Construct a chain of block index entries
+    std::list<CBlockIndex> blockidx;
+    for (auto& [time, bits] : g_mainnet_data) {
```
Comment (knst) [inline src/test/pow_tests.cpp:32]: nit: missing const

```
for (const auto& [time, bits] : g_mainnet_data) {
```

File: src/test/pow_tests.cpp [hunk 2]
```diff
@@ -11,112 +11,40 @@
...
-    CBlockIndex blockIndexPrev22 = CBlockIndex();
-    blockIndexPrev22.nTime = 1408728332;  // Block #123434
-    blockIndexPrev22.nBits = 0x1b10e09e;
-    blockIndexPrev21.pprev = &blockIndexPrev22;
-    CBlockIndex blockIndexPrev23 = CBlockIndex();
-    blockIndexPrev23.nTime = 1408728124;  // Block #123433
-    blockIndexPrev23.nBits = 0x1b104be1;
-    blockIndexPrev22.pprev = &blockIndexPrev23;
+    // Construct a chain of block index entries
+    std::list<CBlockIndex> blockidx;
```
Comment (knst) [inline src/test/pow_tests.cpp:31]: why `std::list`? let's rewrite it with `std::vector`

File: src/test/pow_tests.cpp [hunk 3]
```diff
@@ -11,112 +11,40 @@
 
 BOOST_FIXTURE_TEST_SUITE(pow_tests, BasicTestingSetup)
 
+static const std::vector<std::pair<uint32_t, uint32_t>> g_mainnet_data = {
```
Comment (knst) [inline src/test/pow_tests.cpp:14]: nit: no need to pollute global namespace, it's a local variable for this test.

**Review by knst — COMMENTED** [review]
check 7fddf8fa0cb0dc5002a80c681d17de16291bded3 which fixes mentioned nits

File: src/masternode/sync.cpp [hunk 4]
```diff
@@ -206,7 +206,7 @@ void CMasternodeSync::ProcessTick(const PeerManager& peerman)
                         // Now that the blockchain is synced request the mempool from the connected outbound nodes if possible
                         for (auto pNodeTmp : snap.Nodes()) {
                             bool fRequestedEarlier = m_netfulfilledman.HasFulfilledRequest(pNodeTmp->addr, "mempool-sync");
-                            if (pNodeTmp->nVersion >= 70216 && !pNodeTmp->IsInboundConn() && !fRequestedEarlier && !pNodeTmp->IsBlockRelayOnly()) {
+                            if (!pNodeTmp->IsInboundConn() && !fRequestedEarlier && !pNodeTmp->IsBlockRelayOnly()) {
```
Comment (PastaPastaPasta) [inline src/masternode/sync.cpp:209]: Nope, can't do this like this, we would need to bump minimum protocol version if you wanted, but generally we try to keep lots of support for protocol version. see https://github.com/dashpay/dash/blob/57fb0874ef180cd9aa8b7800a2a0842e2faba65e/src/version.h#L19-L20

@UdjinM6 please confirm my understanding

Comment (UdjinM6) [reply src/masternode/sync.cpp:209]: We keep old proto support mostly for light/mobile clients. They do not have mempool so dropping this condition is ok-ish imo.

Comment (kwvg) [reply src/masternode/sync.cpp:209]: Have bumped `MIN_PEER_PROTO_VERSION` to `70216` on the understanding that clients that don't have a mempool wouldn't care and nodes that do have a mempool would need to be on a far more recent version anyways.

File: src/test/pow_tests.cpp (hunk 1 above)
Comment (kwvg) [reply src/test/pow_tests.cpp]: Resolved in latest push

File: src/test/pow_tests.cpp (hunk 3 above)
Comment (kwvg) [reply src/test/pow_tests.cpp]: Resolved in latest push

File: src/test/pow_tests.cpp (hunk 2 above)
Comment (kwvg) [reply src/test/pow_tests.cpp:31]: I don't see any benefit to using `std::vector` and `std::list` is used by another test ([source](https://github.com/dashpay/dash/blob/d754799580d2914a0dd6388c90100d14feec0428/src/test/skiplist_tests.cpp#L147)), I think it's fine as-is.

File: test/functional/feature_blockfilterindex_prune.py [hunk 5]
```diff
@@ -51,17 +52,17 @@ def run_test(self):
...
         pruneheight_new = self.nodes[0].pruneblockchain(1000)
         assert_greater_than(pruneheight_new, pruneheight)
-        self.stop_node(0, expected_stderr='Warning: You are starting with governance validation disabled. This is expected because you are running a pruned node.')
+        self.stop_node(0, expected_stderr=EXPECTED_STDERR_NO_GOV_PRUNE)
 
         self.log.info("make sure we get an init error when starting the node again with block filters")
         self.nodes[0].assert_start_raises_init_error(
             extra_args=["-fastprune", "-prune=1", "-blockfilterindex=1"],
-            expected_msg="Warning: You are starting with governance validation disabled. This is expected because you are running a pruned node.\nError: basic block filter index best block of the index goes beyond pruned data. Please disable the index or reindex (which will download the whole blockchain again)",
+            expected_msg=EXPECTED_STDERR_NO_GOV_PRUNE + "\nError: basic block filter index best block of the index goes beyond pruned data. Please disable the index or reindex (which will download the whole blockchain again)",
```
Comment (knst) [inline test/functional/feature_blockfilterindex_prune.py:60]: Nit: you can use here f-string to unify with other code:

expected_msg=f"{EXPECTED_STDERR_NO_GOV_PRUNE}\nError: basic block filter index best block of the index goes beyond pruned data. Please disable the index or reindex (which will download the whole blockchain again)",

File: test/functional/test_framework/governance.py [hunk 6]
```diff
@@ -6,6 +6,9 @@
 
 import json
 
+EXPECTED_STDERR_NO_GOV = "Warning: You are starting with governance validation disabled."
+EXPECTED_STDERR_NO_GOV_PRUNE = EXPECTED_STDERR_NO_GOV + " This is expected because you are running a pruned node."
```
Comment (knst) [inline test/functional/test_framework/governance.py:10]: Same here for f-string

File: src/test/pow_tests.cpp (hunk 2 above)
Comment (knst) [reply src/test/pow_tests.cpp:31]: that's a single unit test, not a hot function,so, extra memory allocation is not important for performance. Agree, std::list is fine

**Review by knst — DISMISSED** [review]
utACK aa16a929ac212c8e8a2bd2195cb11d019cad2bd5

**Review by UdjinM6 — DISMISSED** [review]
utACK aa16a929ac212c8e8a2bd2195cb11d019cad2bd5

File: src/masternode/sync.cpp (hunk 4 above)
Comment (PastaPastaPasta) [reply src/masternode/sync.cpp:209]: we should do this in it's own PR imo

File: test/functional/feature_blockfilterindex_prune.py (hunk 5 above)
Comment (kwvg) [reply test/functional/feature_blockfilterindex_prune.py]: Resolved in latest push

File: src/masternode/sync.cpp (hunk 4 above)
Comment (kwvg) [reply src/masternode/sync.cpp]: Change spun off into https://github.com/dashpay/dash/pull/6281

**Review by knst — APPROVED** [review]
re-utACK 4e621037c5553e24decbebefd3a68305b7063b72

**Review by UdjinM6 — APPROVED** [review]
utACK 4e621037c5553e24decbebefd3a68305b7063b72

**Pull request merged**